### PR TITLE
[HZ-989] Implement JSON_OBJECTAGG aggregation for SQL

### PR DESCRIPTION
**Pull request by hakanmemisoglu**
Support for `JSON_OBJECTAGG` aggregation function

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/aggregate/JsonSqlAggregateTest.java [hunk 1]
```diff
@@ -185,4 +229,71 @@ private static String createTable() {
...
+            JsonValue value = actualRowsInMap.get(kv.getKey());
+            JsonObject object = value.asObject();
+            JsonObject object2 = kv.getValue().asObject();
+            if (!(jsonObjectEquals(object, object2))) {
+                throw new AssertionError("Object: " + object + " is not equal to Object: " + object2);
+            }
+        }
+    }
+
+    private boolean jsonObjectEquals(JsonObject obj, JsonObject obj2) {
```
Comment (krzysztofslusarski) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/aggregate/JsonSqlAggregateTest.java:290]: If I read correctly that method allows mixed order of attributes only on the first level, so sucha a JSONs are not equal:
```json
{
    "foo": {
        "bar1" : "bar1",
        "bar2" : "bar2"
    }
}
```
```json
{
    "foo": {
        "bar2" : "bar2",
        "bar1" : "bar1"
    }
}
```

If I'm correct then maybe you should rename the method so it inform about that. But it is just a minor.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/misc/NestingAndCasingExpressionTest.java [hunk 2]
```diff
@@ -50,6 +51,36 @@
 @RunWith(HazelcastSerialClassRunner.class)
 @Category({QuickTest.class, ParallelJVMTest.class})
 public class NestingAndCasingExpressionTest extends ExpressionTestSupport {
+
+    private static final Set<String> EXCEPTIONAL_EXPRESSIONS = new HashSet<>(
```
Comment (krzysztofslusarski) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/misc/NestingAndCasingExpressionTest.java:55]: :heart:

**Review by krzysztofslusarski — APPROVED** [review]
Approve with minors

**Comment by hakanmemisoglu** [comment]
run-lab-run

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/aggregate/JsonSqlAggregateTest.java [hunk 3]
```diff
@@ -185,4 +229,74 @@ private static String createTable() {
...
+                        new String[]{"Bob", "job", "job2"},
+                        new String[]{null, "job", "job1"},
+                        new String[]{null, null, "desc2"},
+                        new String[]{"Bob", "description", null},
+                        new String[]{"Alice", "description", "desc1"})
+        );
+        return name;
+    }
+
+    private void assertRowsWithJsonValues(String sql, Collection<Row> rows) {
```
Comment (Fly-Style) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/aggregate/JsonSqlAggregateTest.java]: Generally, this method looks overcomplicated and it is not universal. E.g., I tried to play with `JSON_OBJECTAGG`, and wrote this test:

```
        @Test
        public void test_sandbox() {
            String name = "map";
            createMapping(name, String.class, int.class);
            IMap<String, Integer> map = instance().getMap(name);
            map.put("A", 1);
            map.put("B", 1);

            assertRowsWithJsonValues(
                    "SELECT JSON_OBJECTAGG(KEY __key VALUE this) FROM " + name,
                    asList(
                            new Row("{\"A\":1,\"B\":1}")
                    )
            );
        }
```
, but I have no success due to assertion errors below. 
Would you like to refactor it and make it more general? Then you can move it to `SqlJsonTestSupport`. WDYT?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/misc/NestingAndCasingExpressionTest.java (hunk 2 above)
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/misc/NestingAndCasingExpressionTest.java:55]: image

**Review by Fly-Style — COMMENTED** [review]
Generally looks correct, some comments above.

**Review by viliam-durina — CHANGES_REQUESTED** [review]
I added a couple of failing tests dealing with null and duplicate keys.

**Review by viliam-durina — APPROVED** [review]

**Pull request merged**